### PR TITLE
refactor: tr_peerIo 

### PR DESCRIPTION
**Pull request by ckerr**
- make tr_peerIo fields private
- simplify socket error handling
- merge similar functions that handled reading from peers
- merge similar functions that handled writing to peers
- use separate callbacks for each peer-io utp event

It's difficult to write unit tests for peer-io, so this would benefit from real-world testers.

### CODE REVIEW

**Comment by Coeur** [comment]
>It's difficult to write unit tests for peer-io, so this would benefit from real-world testers.

Should we put the milestone "Decide after 4.0.0" ;).
Those are more than 800 lines of code to review, where you admit yourself that you're unsure about it.

I wouldn't mind releasing Transmission 4 without extra lengthy refactoring.

File: libtransmission/handshake.cc [hunk 1]
```diff
@@ -771,67 +769,64 @@ ReadState tr_handshake::can_read(tr_peerIo* peer_io, void* vhandshake, size_t* p
...
 
         /* Don't mark a peer as non-µTP unless it's really a connect failure. */
-        if ((errcode == ETIMEDOUT || errcode == ECONNREFUSED) && info)
+        if ((error.code == ETIMEDOUT || error.code == ECONNREFUSED) && info)
         {
             handshake->mediator_->set_utp_failed(info_hash, io->address());
         }
 
-        if (handshake->mediator_->allows_tcp() && io->reconnect() == 0)
+        if (handshake->mediator_->allows_tcp() && io->reconnect())
```
Comment (Coeur) [inline libtransmission/handshake.cc:804]: Is that an intended change (aka, some bug fix)?
Previously we were doing `== 0`, and now it's the same as `!= 0`.

**Review by Coeur — CHANGES_REQUESTED** [review]
After a few minutes, crash on:
```
TR_ASSERT_MSG(false, "unsupported peer socket type");
```

```
2022-12-16 09:38:39.596237+0800 Transmission[12016:2054821] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'assertion failed: unsupported peer socket type (/Users/User/Developer/transmission/libtransmission/peer-io.cc:279)'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000019b7173f8 __exceptionPreprocess + 176
	1   libobjc.A.dylib                     0x000000019b262ea8 objc_exception_throw + 60
	2   CoreFoundation                      0x000000019b71729c +[NSException exceptionWithName:reason:userInfo:] + 0
	3   Transmission                        0x00000001058fcaf0 _Z16tr_assert_reportNSt3__117basic_string_viewIcNS_11char_traitsIcEEEEiS3_ + 1656
	4   Transmission                        0x00000001056a8bb4 _ZN9tr_peerIo10set_socketE14tr_peer_socket + 1088
	5   Transmission                        0x00000001056ab0f4 _ZN9tr_peerIo12new_outgoingEP10tr_sessionP12tr_bandwidthRK10tr_address7tr_portRKNSt3__15arrayISt4byteLm20EEEbb + 2580
	6   Transmission                        0x00000001056f1080 _ZN15connect_helpers18initiateConnectionEP10tr_peerMgrP8tr_swarmR9peer_atom + 3344
	7   Transmission                        0x00000001056ee9a0 _ZN10tr_peerMgr22makeNewPeerConnectionsEm + 956
	8   Transmission                        0x00000001056ede80 _ZN10tr_peerMgr14reconnectPulseEv + 1672
	9   Transmission                        0x00000001056f0220 _ZN10tr_peerMgr14bandwidthPulseEv + 1420
	10  Transmission                        0x00000001056fb0d0 _ZZN10tr_peerMgrC1EP10tr_sessionENKUlvE_clEv + 80
	11  Transmission                        0x00000001056fb060 _ZNSt3__1L8__invokeIRZN10tr_peerMgrC1EP10tr_sessionEUlvE_JEEEDTclscT_fp_spscT0_fp0_EEOS6_DpOS7_ + 24
	12  Transmission                        0x00000001056fb018 _ZNSt3__128__invoke_void_return_wrapperIvLb1EE6__callIJRZN10tr_peerMgrC1EP10tr_sessionEUlvE_EEEvDpOT_ + 28
	13  Transmission                        0x00000001056faff0 _ZNSt3__110__function12__alloc_funcIZN10tr_peerMgrC1EP10tr_sessionEUlvE_NS_9allocatorIS5_EEFvvEEclEv + 28
	14  Transmission                        0x00000001056f6f38 _ZNSt3__110__function6__funcIZN10tr_peerMgrC1EP10tr_sessionEUlvE_NS_9allocatorIS5_EEFvvEEclEv + 28
	15  Transmission                        0x0000000105656128 _ZNKSt3__110__function12__value_funcIFvvEEclEv + 260
	16  Transmission                        0x0000000105655150 _ZNKSt3__18functionIFvvEEclEv + 24
	17  Transmission                        0x0000000105bf3888 _ZN15libtransmission7EvTimer11handleTimerEv + 684
	18  Transmission                        0x0000000105bf26bc _ZN15libtransmission7EvTimer7onTimerEisPv + 32
	19  Transmission                        0x0000000105cb87e4 event_persist_closure + 3924
	20  Transmission                        0x0000000105cb5e70 event_process_active_single_queue + 2468
	21  Transmission                        0x0000000105ca4264 event_process_active + 1760
	22  Transmission                        0x0000000105ca05b8 event_base_loop + 3160
	23  Transmission                        0x000000010565367c _ZN22tr_session_thread_impl17sessionThreadFuncEP10event_base + 688
	24  Transmission                        0x0000000105659288 _ZNSt3__1L8__invokeIM22tr_session_thread_implFvP10event_baseEPS1_JS3_EvEEDTcldsdescT0_fp0_fp_spscT1_fp1_EEOT_OS7_DpOS8_ + 320
	25  Transmission                        0x0000000105659128 _ZNSt3__1L16__thread_executeINS_10unique_ptrINS_15__thread_structENS_14default_deleteIS2_EEEEM22tr_session_thread_implFvP10event_baseEJPS6_S8_EJLm2ELm3EEEEvRNS_5tupleIJT_T0_DpT1_EEENS_15__tuple_indicesIJXspT2_EEEE + 76
	26  Transmission                        0x0000000105657d84 _ZNSt3__1L14__thread_proxyINS_5tupleIJNS_10unique_ptrINS_15__thread_structENS_14default_deleteIS3_EEEEM22tr_session_thread_implFvP10event_baseEPS7_S9_EEEEEPvSE_ + 340
	27  libsystem_pthread.dylib             0x0000000106ef15d4 _pthread_start + 148
	28  libsystem_pthread.dylib             0x0000000106efba88 thread_start + 8
)
2022-12-16 09:38:39.616210+0800 Transmission[12016:2054821] [General] An uncaught exception was raised
2022-12-16 09:38:39.616606+0800 Transmission[12016:2054821] [General] assertion failed: unsupported peer socket type (/Users/User/Developer/transmission/libtransmission/peer-io.cc:279)
2022-12-16 09:38:39.616812+0800 Transmission[12016:2054821] [General] (
	0   CoreFoundation                      0x000000019b7173f8 __exceptionPreprocess + 176
	1   libobjc.A.dylib                     0x000000019b262ea8 objc_exception_throw + 60
	2   CoreFoundation                      0x000000019b71729c +[NSException exceptionWithName:reason:userInfo:] + 0
	3   Transmission                        0x00000001058fcaf0 _Z16tr_assert_reportNSt3__117basic_string_viewIcNS_11char_traitsIcEEEEiS3_ + 1656
	4   Transmission                        0x00000001056a8bb4 _ZN9tr_peerIo10set_socketE14tr_peer_socket + 1088
	5   Transmission                        0x00000001056ab0f4 _ZN9tr_peerIo12new_outgoingEP10tr_sessionP12tr_bandwidthRK10tr_address7tr_portRKNSt3__15arrayISt4byteLm20EEEbb + 2580
	6   Transmission                        0x00000001056f1080 _ZN15connect_helpers18initiateConnectionEP10tr_peerMgrP8tr_swarmR9peer_atom + 3344
	7   Transmission                        0x00000001056ee9a0 _ZN10tr_peerMgr22makeNewPeerConnectionsEm + 956
	8   Transmission                        0x00000001056ede80 _ZN10tr_peerMgr14reconnectPulseEv + 1672
	9   Transmission                        0x00000001056f0220 _ZN10tr_peerMgr14bandwidthPulseEv + 1420
	10  Transmission                        0x00000001056fb0d0 _ZZN10tr_peerMgrC1EP10tr_sessionENKUlvE_clEv + 80
	11  Transmission                        0x00000001056fb060 _ZNSt3__1L8__invokeIRZN10tr_peerMgrC1EP10tr_sessionEUlvE_JEEEDTclscT_fp_spscT0_fp0_EEOS6_DpOS7_ + 24
	12  Transmission                        0x00000001056fb018 _ZNSt3__128__invoke_void_return_wrapperIvLb1EE6__callIJRZN10tr_peerMgrC1EP10tr_sessionEUlvE_EEEvDpOT_ + 28
	13  Transmission                        0x00000001056faff0 _ZNSt3__110__function12__alloc_funcIZN10tr_peerMgrC1EP10tr_sessionEUlvE_NS_9allocatorIS5_EEFvvEEclEv + 28
	14  Transmission                        0x00000001056f6f38 _ZNSt3__110__function6__funcIZN10tr_peerMgrC1EP10tr_sessionEUlvE_NS_9allocatorIS5_EEFvvEEclEv + 28
	15  Transmission                        0x0000000105656128 _ZNKSt3__110__function12__value_funcIFvvEEclEv + 260
	16  Transmission                        0x0000000105655150 _ZNKSt3__18functionIFvvEEclEv + 24
	17  Transmission                        0x0000000105bf3888 _ZN15libtransmission7EvTimer11handleTimerEv + 684
	18  Transmission                        0x0000000105bf26bc _ZN15libtransmission7EvTimer7onTimerEisPv + 32
	19  Transmission                        0x0000000105cb87e4 event_persist_closure + 3924
	20  Transmission                        0x0000000105cb5e70 event_process_active_single_queue + 2468
	21  Transmission                        0x0000000105ca4264 event_process_active + 1760
	22  Transmission                        0x0000000105ca05b8 event_base_loop + 3160
	23  Transmission                        0x000000010565367c _ZN22tr_session_thread_impl17sessionThreadFuncEP10event_base + 688
	24  Transmission                        0x0000000105659288 _ZNSt3__1L8__invokeIM22tr_session_thread_implFvP10event_baseEPS1_JS3_EvEEDTcldsdescT0_fp0_fp_spscT1_fp1_EEOT_OS7_DpOS8_ + 320
	25  Transmission                        0x0000000105659128 _ZNSt3__1L16__thread_executeINS_10unique_ptrINS_15__thread_structENS_14default_deleteIS2_EEEEM22tr_session_thread_implFvP10event_baseEJPS6_S8_EJLm2ELm3EEEEvRNS_5tupleIJT_T0_DpT1_EEENS_15__tuple_indicesIJXspT2_EEEE + 76
	26  Transmission                        0x0000000105657d84 _ZNSt3__1L14__thread_proxyINS_5tupleIJNS_10unique_ptrINS_15__thread_structENS_14default_deleteIS3_EEEEM22tr_session_thread_implFvP10event_baseEPS7_S9_EEEEEPvSE_ + 340
	27  libsystem_pthread.dylib             0x0000000106ef15d4 _pthread_start + 148
	28  libsystem_pthread.dylib             0x0000000106efba88 thread_start + 8
)
```

**Comment by Coeur** [comment]
Crash reproduced a second time, after 5-10 minutes.
Capture d’écran 2022-12-16 à 10 13 03

**Comment by Coeur** [comment]
Note also that a *lot* of those changes are just renaming: could you make:
1. one PR that just do the safe renaming part and trivial changes (like inverting the `reconnect()` returned value)
2. we merge it
3. and then another PR for the remaining unsafe changes, with much fewer lines modified and easier to analyze for the cause of potential regressions

File: libtransmission/handshake.cc (hunk 1 above)
Comment (ckerr) [reply libtransmission/handshake.cc:804]: Yes, https://github.com/transmission/transmission/pull/4372/commits/0fed1f43436db1981cf8467261498adc5407255a -- reconnect now returns a bool. Previously it used the pre-bool-C-ism of returning 0 on success, -1 on error.

**Review by Coeur — APPROVED** [review]
I didn't do an in-depth code review this time, only surface review.
I did testing, downloading, pausing, quitting. Seems working OK.

**Comment by ckerr** [comment]
> Should we put the milestone "Decide after 4.0.0" ;). Those are more than 800 lines of code to review, where you admit yourself that you're unsure about it. I wouldn't mind releasing Transmission 4 without extra lengthy refactoring.

I get your point, but I think it's better to get all of this out of the way while we're in the beta cycle. The main refactors I have in mind post-4.0.0 are the ones for the 5 milestone, e.g. full BitTorrent v2 support and nonblocking disk IO.

**Comment by GaryElshaw** [comment]
> full BitTorrent v2 support and nonblocking disk IO

Yay!

**Pull request merged**